Add support for more fields

romaninsh opened the pull request:
implement types using DBAL first - https://github.com/atk4/data/pull/727

Implements separate classes for Number, Percent and Rating. Also includes Money for a basic implementation.

romaninsh committed:
added number field support

romaninsh committed:
Apply fixes from StyleCI

romaninsh committed:
Merge remote-tracking branch 'origin/feature/epic-field-refactor' into feature/number-field

romaninsh committed:
Merge branch 'feature/number-field' of github.com:atk4/data into feature/number-field

DarkSide666 committed:
Update Money.php

DarkSide666 committed:
Update Money.php

romaninsh committed:
Merge branch 'develop' into feature/number-field

DarkSide666 committed:
fix Boolean field

DarkSide666 committed:
Apply fixes from StyleCI

DarkSide666 committed:
implement Field/Numeric

DarkSide666 committed:
Apply fixes from StyleCI

DarkSide666 committed:
Implement Field/Integer and Field/Money

DarkSide666 committed:
Apply fixes from StyleCI

romaninsh commented on src/Field/Integer.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Specify how many decimal numbers should be saved.
+     */
+    public $decimalNumbers = 0;
```
we commonly use $decimal_number, no caps.. i think.

DarkSide666 replied:
addressed

DarkSide666 committed:
fix tests

DarkSide666 committed:
Apply fixes from StyleCI

romaninsh commented on src/Field/Integer.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Enable number rounding. If true will round number, otherwise will round it down (trim).
+     */
+    public $enableRounding = false;
```
already defined in a parent.

DarkSide666 replied:
renamed to enable_rounding and it's defined here because it changes default value (to false) for integer.

romaninsh commented on src/Persistence/Array_.php:
```diff
@@ -56,7 +56,10 @@ public function add($m, $defaults = [])
+                //$f->type = 'integer';
+                // have to re-add field to use proper Field class
+                $m->removeField($m->id_field);
+                $m->addField($m->id_field, ['type'=>'integer']);
```
probably could be decided earlier, before field is added.

DarkSide666 replied:
Not really because ID field is added in Model automatically even before Model is added into Persistence.

romaninsh replied:
ok, lets have id field seed then, which is by default id_field_seed = ['Integer'] but could also be something else.

DarkSide666 replied:
Added id field seed property in Model class and removed $type property dependency from Array Persistence class.

romaninsh reviewed (commented):
WIP..

DarkSide666 committed:
address Romans comments

DarkSide666 committed:
simplify

DarkSide666 committed:
add Text and ShortText fields

DarkSide666 committed:
shorttext

romaninsh commented on src/Field.php:
```diff
@@ -269,19 +275,9 @@ public function normalize($value)
-                throw new ValidationException([$this->name => 'Must not be empty']);
-            }
-            break;
+            throw new Exception(['Use Field\Text for type=text', 'this'=>$this]);
```
didnt' we agree on the meeting to keep backwards compatibility ?

if we accept that, it will break code in a very "python" way (no steady update path)

DarkSide666 replied:
It should be BC so far. Derived field classes don't call parent::normalize() so this switch is not used at all. Theoretically all this normalize method in Field class can be removed, but not 100% yet. That's why I left it there for cases when Field->type === null (some custom fields).

romaninsh commented on src/Field/ShortText.php:
```diff
@@ -0,0 +1,42 @@
+ *
+ * Most of the time this is most basic type you will use.
+ */
+class ShortText extends Text
```
I suggest we have two classes:

Line (may be a short or a long line, but enter is not permitted)
Text (may be a short or a long text, will contain some paragraphs)

DarkSide666 replied:
Addressed.
Simply renaming ShortText class to Line does this :)
No more need for _short_ vs _long_. Limit can be defined by max_length property in both of these classes.

DarkSide666 committed:
fix comments

DarkSide666 committed:
Apply fixes from StyleCI

DarkSide666 committed:
refactor datetime, date, time fields

DarkSide666 committed:
Apply fixes from StyleCI

DarkSide666 committed:
refactor array and object fields

DarkSide666 committed:
Apply fixes from StyleCI

DarkSide666 committed:
add canBeNull, canBeEmpty methods

DarkSide666 committed:
types

romaninsh commented on src/Field.php:
```diff
@@ -235,195 +217,79 @@ public function __construct($defaults = [])
-                throw new ValidationException([$this->name => 'Must use scalar value']);
+        // only string type fields can use empty string as legit value, for all
+        // other field types empty value is the same as no-value, nothing or null
+        if ($f->type && $f->type != 'string' && $value === '') {
```
probably incorrect.......

DarkSide666 replied:
what exactly?

romaninsh commented on src/Field.php:
```diff
@@ -235,195 +217,79 @@ public function __construct($defaults = [])
-            if (!$this->owner->strict_types) {
-                return $value;
+        if ($value === null) {
+            if ($this->required) {
```
rename into not_null

romaninsh commented on src/Field.php:
```diff
@@ -235,195 +217,79 @@ public function __construct($defaults = [])
-        } catch (Exception $e) {
-            $e->addMoreInfo('field', $this);
+        // validate scalar values
+        if (in_array($f->type, ['string', 'text', 'integer', 'money', 'float']) && !is_scalar($value)) {
```
we don't care about type, remove.

romaninsh commented on src/Field.php:
```diff
@@ -235,195 +217,79 @@ public function __construct($defaults = [])
-            throw $e;
+        // normalize
+        // @TODO remove this block in future - it's useless
+        switch ($f->type) {
```
switch not needed. remove.

romaninsh commented on src/Field.php:
```diff
@@ -475,49 +341,66 @@ public function useAlias()
+     */
+    public function canBeNull() : bool
+    {
+        return $this->mandatory === false;
```
return !$this->not_null

romaninsh commented on src/Field.php:
```diff
@@ -475,49 +341,66 @@ public function useAlias()
+     */
+    public function canBeEmpty() : bool
+    {
+        return $this->mandatory === false && $this->required === false;
```
return !$this->['ui']['required'] ?? true;

abbadon1334 replied:
@romaninsh '$this->['ui']['required']' <--- i lost some meeting but why check for UI required in place of required ?

romaninsh commented on src/Field/Array_.php:
```diff
@@ -0,0 +1,67 @@
+    {
+        if ($value === null || $value === '') {
+            if ($this->required) {
+                throw new ValidationException([$this->name => 'Must not be null or empty']);
```
shouldn't we call parent?

DarkSide666 replied:
I am avoiding calling parent normalize because was trying to get rid of it at first and then see if a lot of field classes will actually duplicate same code over and over again, then move it back to parent class normalize method.

romaninsh commented on src/Field.php:
```diff
@@ -222,6 +202,8 @@ public function __construct($defaults = [])
     }
```
set($this->default_value)

romaninsh commented on src/Field/Boolean.php:
```diff
@@ -55,36 +58,42 @@ public function init()
-        if (is_numeric($value)) {
-            return (bool) $value;
+        // if value required, then only valueTrue is allowed
+        if ($this->required && $value !== true) {
```
technically should be now job of checkbox, but we can leave it for now..

romaninsh commented on src/Field/DateTime.php:
```diff
@@ -0,0 +1,105 @@
+     * @var array
+     */
+    public $persistence = [
+        'format'   => null, // for date it can be 'Y-m-d', for datetime - 'Y-m-d H:i:s' etc.
```
why not use Y-m-d as default ?

romaninsh commented on src/Field/DateTime.php:
```diff
@@ -0,0 +1,105 @@
+    {
+        if ($value === null) {
+            if ($this->mandatory || $this->required) {
+                throw new ValidationException([$this->name => 'Must not be null']);
```
not_null

DarkSide666 committed:
work in progress

romaninsh commented on src/Field/Money.php:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * @var int Specify how many decimal numbers should be saved.
+     */
+    public $decimal_numbers = 4;
```
normally that's 2

DarkSide666 replied:
it's 4 currently so I thought we should have the same (for BC), but normally it should be 2 in most cases.

DarkSide666 replied:
addressed

DarkSide666 committed:
Apply fixes from StyleCI

romaninsh commented on src/Field/Numeric.php:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return float
+     */
+    protected function round_up(float $n, int $p) : float
```
roundUp

DarkSide666 replied:
addressed

romaninsh commented on src/Field/Numeric.php:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return float
+     */
+    protected function round_down(float $n, int $p) : float
```
roundDown

DarkSide666 replied:
addressed

romaninsh commented on src/Field/Object_.php:
```diff
@@ -0,0 +1,67 @@
+    public function normalize($value)
+    {
+        if ($value === null || $value === '') {
+            if ($this->required) {
```
parent call? not_null..

abbadon1334 replied:
not_null will be another extra property we already have mandatory/system/required that are still hard to understand at first, i suggest to hardcode in a parent call or in a specific protected method like : 

```php
protected function notNull($value) : void {
    if ((null === $value || $value === '') && $this->required) {
        throw new ValidationException([$this->name => 'Must not be null or empty']);
    }
}
```

romaninsh commented on src/Field/Object_.php:
```diff
@@ -0,0 +1,67 @@
+    {
+        if ($value === null || $value === '') {
+            if ($this->required) {
+                throw new ValidationException([$this->name => 'Must not be null or empty']);
```
how can object be empty ?

DarkSide666 replied:
addressed

romaninsh commented on src/Field/Object_.php:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        if (is_string($value) && $this->owner && $this->owner->persistence) {
+            $value = $this->owner->persistence->jsonDecode($this, $value, false);
```
is that needed ?

romaninsh commented on src/Field/Object_.php:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return string
+     */
+    public function toString($value = null) : ?string
```
use native serialization

DarkSide666 replied:
why? jsonEncode in persistence class is native, but _better_ method because it treats exception cases better.

romaninsh commented on src/Field/_Percent.php:
```diff
@@ -0,0 +1,19 @@
+    /**
+     * @var int IMPORTANT: set to 2+precision, since percentage is stored as a 0 .. 1
+     */
+    public $decimalNumbers = 2;
```
decimal_numbers

DarkSide666 replied:
addressed

romaninsh commented on src/Model.php:
```diff
@@ -482,8 +482,19 @@ public function fieldFactory($seed = [])
     protected $typeToFieldSeed = [
-        'boolean' => ['Boolean'],
+        'boolean'  => ['Boolean'],
+        'float'    => ['Numeric'],
```
'float' => ['LegacyField', 'type'=>'float'] 

etc

DarkSide666 committed:
add id field seed property, remove type property usage from Array Persistence class

DarkSide666 committed:
Apply fixes from StyleCI

DarkSide666 committed:
Rename ShortText to Line

DarkSide666 committed:
Merge branch 'develop' into feature/number-field

DarkSide666 committed:
require more extensions

DarkSide666 committed:
rename round methods

DarkSide666 committed:
fix property name

DarkSide666 committed:
fix Percent class

DarkSide666 committed:
address few comments

DarkSide666 committed:
fix

DarkSide666 committed:
Merge branch 'develop' into feature/number-field

DarkSide666 committed:
fix tests

DarkSide666 committed:
Merge branch 'develop' into feature/number-field

abbadon1334 reviewed (changes requested):
LGTM, i agree to hardcode in some way the notNull exception

georgehristov committed:
use property to define field seed properties (#509)

* badge

* Feature/set custom edit exec button (#490)

* fix one test

* Setting current dependencies

* use custom exec button on edit action

* Update Model.php

* remove class, use seed

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* model default add field property (#454)

* use the object default add field property

* no need for double brackets.

* force tests

Co-authored-by: Romans Malinovskis <me@nearly.guru>
Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* Fix empty array condition (#498)

* more tests

* Apply fixes from StyleCI

* better implementation for addFields() and test cases (#499)

* better implementation for addFields() and test cases

* Apply fixes from StyleCI

* use shorthand method

* simplify code

* implement `exprNow()` method. Rely on dsql PR.

* Apply fixes from StyleCI

* pass

* add docs

Co-authored-by: Romans Malinovskis <me@nearly.guru>

* Fix/ Small Action related fix (#502)

* Fix/ Small Action related fix

- Set Edit action exec button labe lto 'Save' by default
 - Set Edit action exec button colot to blue by default
- Set default description for Add action in model
- Add method to retreive model from Action

* Apply fixes from StyleCI

* fix comment

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* use property to define field seed properties

* update field test

* introduce expandable field type registry

* update test to new properties

* revert type setting

* add shorthand int type declaration

* refactor variable names for clarity and consistency

* introduce Field::getTypecaster and Field::getSerializer methods

* introduce registering field seeds from array

* chore: refactor variable names for consistency

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>
Co-authored-by: Mimo <ddobchev@gmail.com>
Co-authored-by: Romans Malinovskis <me@nearly.guru>
Co-authored-by: Alain Belair <belair.alain@gmail.com>

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
chore: merge develop into feature/number-field  (#528)

* badge

* Feature/set custom edit exec button (#490)

* fix one test

* Setting current dependencies

* use custom exec button on edit action

* Update Model.php

* remove class, use seed

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* model default add field property (#454)

* use the object default add field property

* no need for double brackets.

* force tests

Co-authored-by: Romans Malinovskis <me@nearly.guru>
Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* Fix empty array condition (#498)

* more tests

* Apply fixes from StyleCI

* better implementation for addFields() and test cases (#499)

* better implementation for addFields() and test cases

* Apply fixes from StyleCI

* use shorthand method

* simplify code

* implement `exprNow()` method. Rely on dsql PR.

* Apply fixes from StyleCI

* pass

* add docs

Co-authored-by: Romans Malinovskis <me@nearly.guru>

* Fix/ Small Action related fix (#502)

* Fix/ Small Action related fix

- Set Edit action exec button labe lto 'Save' by default
 - Set Edit action exec button colot to blue by default
- Set default description for Add action in model
- Add method to retreive model from Action

* Apply fixes from StyleCI

* fix comment

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* update multiple delete example (#503)

* now supports multiple filter options in getFields()

* Apply fixes from StyleCI

* Accept any DateTimeInterface impl. for datetime (#505)

* Accept any DateTimeInterface impl. for datetime

* Fix DateTime::getTimezone method name

* Fix setTimezone for any instance of DateTimeInterface

* Add microseconds persistence support for datetime/time types + fix normalization/cloning issue (#504)

* Fix datetime normalization cloning

* Add microseconds persistence support for datetime/time types

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>

* include comment about hooks and example how to execute them (#510)

* update composer

* Update release-drafter.yml

* Update bundler.yml

* Fix hasOne relation seed processing. It should replace not merge. (#512)

* New method $model->getTitles() (#513)

* Implement `getTitles()` method

* Apply fixes from StyleCI

* Update unit-tests.yml

* migration to migrator

* rename addHook to onHook (#514)

* fix #944 (#516)

* allow dots in table names, fix #515, fix #517

* Apply fixes from StyleCI

* Do not fail-fast PHP test matrix (#522)

* Simplify code (#519)

* Update release-drafter.yml

* Update release-drafter.yml

* Fix hook trait usage (#525)

* Hook args must be an array

* Fix hook onHook() usage

* Apply fixes from StyleCI

* initialization of fields and handling of expressions

* Apply fixes from StyleCI

* introduce hasPersistence and retrieval of setting

* Apply fixes from StyleCI

Co-authored-by: Imants Horsts <DarkSide666@users.noreply.github.com>
Co-authored-by: Mimo <ddobchev@gmail.com>
Co-authored-by: Romans Malinovskis <me@nearly.guru>
Co-authored-by: Alain Belair <belair.alain@gmail.com>
Co-authored-by: DarkSide <imants.horsts@inbox.lv>
Co-authored-by: Michael Voříšek <mvorisek@mvorisek.cz>

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
Merge branch 'develop' into feature/number-field

# Conflicts:
#	composer.json
#	locale/en/atk.php
#	locale/it/atk.php
#	locale/ru/atk.php
#	src/Field.php
#	src/Field/Boolean.php
#	src/Field/Callback.php
#	src/Field_SQL.php
#	src/Model.php
#	src/Persistence.php
#	src/Persistence/Array_.php
#	src/Persistence/SQL.php
#	src/Reference/HasOne.php
#	src/Reference/HasOne_SQL.php
#	src/Util/DeepCopy.php
#	tests/FieldTest.php
#	tests/FieldTypesTest.php
#	tests/RandomTest.php

romaninsh committed:
Apply fixes from StyleCI

mvorisek commented:
A lot of changes are already integrated, different fields itself should be fixed in #727, closing this one.
